Add (basic) UTF-8 support in the `stringToPDFString` helper function (issue 14449)

Snuffleupagus opened the pull request:
This patch implements this by looking for the UTF-8 BOM, i.e. `\xEF\xBB\xBF`, in order to determine the encoding.[1]
The actual conversion is done using the `TextDecoder` interface, which should be available in all environments/browsers that we support; please see https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder#browser_compatibility

Fixes #14449

---
[1] Assuming that everything lacking a UTF-16 BOM would have to be UTF-8 encoded really doesn't seem correct.

Snuffleupagus committed:
Add (basic) UTF-8 support in the `stringToPDFString` helper function (issue 14449)

This patch implements this by looking for the UTF-8 BOM, i.e. `\xEF\xBB\xBF`, in order to determine the encoding.[1]
The actual conversion is done using the `TextDecoder` interface, which should be available in all environments/browsers that we support; please see https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder#browser_compatibility

---
[1] Assuming that everything lacking a UTF-16 BOM would have to be UTF-8 encoded really doesn't seem correct.

calixteman commented:
According to:
https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/encoding
it should be possible to use it to for utf-16 (le/be) which would be more consistent.

Snuffleupagus committed:
Re-factor the `stringToPDFString` helper function for UTF-16 strings

This patch changes the function to instead utilize the `TextDecoder` for both kinds of UTF-16 BOM strings.

Snuffleupagus commented:
> According to:
> https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/encoding
> it should be possible to use it to for utf-16 (le/be) which would be more consistent.

Sure, let's try that instead; however I did that change in a *separate* commit since it's not directly related to the issue at hand.

Snuffleupagus commented:
/botio test

calixteman reviewed (approved):
Looks good overall.
I wonder if it could be interesting to add a "fast" path with a single check on `str[0]` (something like `if (str[0] >= "\xEF") { /* check for encoding */ }` in order to avoid 4 checks before reaching the general case (note that I've absolutely no opinion, I'm just wondering).

Snuffleupagus commented:
> Looks good overall. I wonder if it could be interesting to add a "fast" path with a single check on `str[0]` (something like `if (str[0] >= "\xEF") { /* check for encoding */ }` in order to avoid 4 checks before reaching the general case (note that I've absolutely no opinion, I'm just wondering).

Well, it really cannot hurt so why not :-)

/botio unittest
